src/analysis/processing/qgsalgorithmgenerateelevationprofile.cpp

```diff
@@ -0,0 +1,323 @@
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsalgorithmgenerateelevationprofile.h"
+
+#include "qgis.h"
+#include "qgsabstractprofilesource.h"
+#include "qgstextformat.h"
+#include "qgsfillsymbol.h"
+#include "qgsfillsymbollayer.h"
+#include "qgslinesymbol.h"
+#include "qgslinesymbollayer.h"
+#include "qgsplot.h"
+#include "qgsprofilerequest.h"
+#include "qgsterrainprovider.h"
+#include "qgscurve.h"
+
+///@cond PRIVATE
+
+class QgsAlgorithmElevationProfilePlotItem: public Qgs2DPlot
+{
+  public:
+
+    explicit QgsAlgorithmElevationProfilePlotItem( int width, int height, int dpi )
+      : mDpi( dpi )
+    {
+      setYMinimum( 0 );
+      setYMaximum( 10 );
+      setSize( QSizeF( width, height ) );
+    }
+
+    void setRenderer( QgsProfilePlotRenderer *renderer )
+    {
+      mRenderer = renderer;
+    }
+
+    QRectF plotArea()
+    {
+      if ( !mPlotArea.isNull() )
+      {
+        return mPlotArea;
+      }
+
+      // calculate plot area
+      QgsRenderContext context;
+      context.setScaleFactor( mDpi / 25.4 );
+
+      calculateOptimisedIntervals( context );
+      mPlotArea = interiorPlotArea( context );
+      return mPlotArea;
+    }
+
+    void renderContent( QgsRenderContext &rc, const QRectF &plotArea ) override
+    {
+      mPlotArea = plotArea;
+
+      if ( !mRenderer )
+        return;
+
+      rc.painter()->translate( mPlotArea.left(), mPlotArea.top() );
+      const QStringList sourceIds = mRenderer->sourceIds();
+      for ( const QString &source : sourceIds )
+      {
+        mRenderer->render( rc, mPlotArea.width(), mPlotArea.height(), xMinimum(), xMaximum(), yMinimum(), yMaximum(), source );
+      }
+      rc.painter()->translate( -mPlotArea.left(), -mPlotArea.top() );
+    }
+
+  private:
+
+    int mDpi = 96;
+    QRectF mPlotArea;
+    QgsProfilePlotRenderer *mRenderer = nullptr;
+};
+
+void QgsGenerateElevationProfileAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterGeometry( QStringLiteral( "CURVE" ), QObject::tr( "Profile curve" ), QVariant(), false,  QList<int>() << static_cast<int>( Qgis::GeometryType::Line ) ) );
+  addParameter( new QgsProcessingParameterCrs( QStringLiteral( "CURVE_CRS" ), QObject::tr( "Profile curve CRS" ), QVariant(), false ) );
```

Comment:
here we should avoid the explicit crs parameter and instead use parameterAsGeometryCrs with the CURVE parameter instead